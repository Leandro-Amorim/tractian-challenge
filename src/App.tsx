import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import PageHeader from "./components/PageHeader";
import MainContent from "./components/MainContent";
import { RecoilRoot } from "recoil";

export default function App() {

	const queryClient = new QueryClient()

	return (
		<RecoilRoot>
			<QueryClientProvider client={queryClient}>
				<div className="min-h-screen bg-background flex flex-col">
					<PageHeader />
					<MainContent />
				</div>
			</QueryClientProvider>
		</RecoilRoot>
	)
}