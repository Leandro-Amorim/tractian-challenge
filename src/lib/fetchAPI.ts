export default async function fetchAPI<T = unknown>(input: string) {
	const response = await fetch(`${import.meta.env.VITE_BASE_API_URL}${input}`, {
		method: 'GET',
	});

	if (!response.ok) {
		throw new Error('Network fail');
	}
	return response.json() as Promise<T>;
}