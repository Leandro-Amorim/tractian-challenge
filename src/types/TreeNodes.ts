export type NodeType = 'location' | 'asset' | 'component';

interface IBasicNode {
	id: string,
	name: string,
	level: number,
	parentId: string | null,
	hasChildren: boolean,
	children: ITreeNode[],
	type: NodeType
}

export interface ILocationNode extends IBasicNode {
	type: 'location',
}

export interface IAssetNode extends IBasicNode {
	type: 'asset',
}

export interface IComponentNode extends IBasicNode {
	type: 'component',
	
	sensorId: string,
	sensorType: 'energy' | 'vibration',
	status: 'operating' | 'alert',
	gatewayId: string,
}

export type ITreeNode = ILocationNode | IAssetNode | IComponentNode;