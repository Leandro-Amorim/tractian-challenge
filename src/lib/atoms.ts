import { atom } from "recoil";
import { ISearch } from "../types/General";
import { IComponentNode } from "../types/TreeNodes";

export const selectedCompanyState = atom({
	key: 'selectedCompanyState', default: {
		id: '',
		name: '',
	}
})

export const searchState = atom({
	key: 'searchState', default: {
		query: '',
		energySensor: false as boolean,
		alertState: false as boolean,
		collapsedIds: [] as string[],
	} satisfies ISearch
})

export const selectedState = atom({
	key: 'selectedState', default: null as IComponentNode | null
})