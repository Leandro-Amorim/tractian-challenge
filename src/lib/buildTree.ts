import { APIAsset, APILocation } from "../types/API";
import { IAssetNode, IComponentNode, ILocationNode, ITreeNode } from "../types/TreeNodes";

export default function buildTree({ locations, assets }: { locations: APILocation[], assets: APIAsset[] }) {

	const result: ITreeNode[] = [];
	const nodeMap: Record<string, ITreeNode> = {};

	for (const location of locations) {
		nodeMap[location.id] = {
			id: location.id,
			name: location.name,
			type: 'location',
			parentId: location.parentId ?? null,
			hasChildren: false,
			children: [],
			level: 0
		} satisfies ILocationNode;
	}

	for (const asset of assets) {

		if (asset.sensorType) {
			nodeMap[asset.id] = {
				id: asset.id,
				name: asset.name,
				type: 'component',
				sensorId: asset.sensorId,
				sensorType: asset.sensorType,
				status: asset.status,
				gatewayId: asset.gatewayId,
				parentId: asset.parentId ?? asset.locationId ?? null,
				hasChildren: false,
				children: [],
				level: 0
			} satisfies IComponentNode;
		}
		else {
			nodeMap[asset.id] = {
				id: asset.id,
				name: asset.name,
				type: 'asset',
				parentId: asset.parentId ?? asset.locationId ?? null,
				hasChildren: false,
				children: [],
				level: 0
			} satisfies IAssetNode;
		}
	}

	for (const id of Object.keys(nodeMap)) {

		const node = nodeMap[id];
		const parentId = node.parentId;
		
		if (parentId) {
			nodeMap[parentId].children.push(node);
			nodeMap[parentId].hasChildren = true;
		}
		else {
			result.push(node);
		}
	}

	for (const node of result) {
		setLevel(node, 0);
	}

	return result;
}

function setLevel(node: ITreeNode, level: number) {
	node.level = level;
	for (const child of node.children) {
		setLevel(child, level + 1);
	}
}