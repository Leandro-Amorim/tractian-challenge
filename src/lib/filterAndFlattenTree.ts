import { ISearch } from "../types/General";
import { ITreeNode } from "../types/TreeNodes";
import { isCollapsed } from "./utils";

export default function filterAndFlattenTree(tree: ITreeNode[], query: ISearch) {

	const filteredNodes = tree.map(node => filterNode(node, query));
	const flattenedList: ITreeNode[] = [];

	for (const node of filteredNodes) {
		flattenTree(node, flattenedList, query);
	}
	return flattenedList;
}

export function filterNode(node: ITreeNode, query: ISearch, fatherShowing: boolean = false) {

	const nameCheck = fatherShowing || node.name.toLowerCase().includes(query.query.toLowerCase());

	let passing = nameCheck;

	if (query.energySensor) {
		passing = passing && (node.type === 'component' && node.sensorType === 'energy');
	}

	if (query.alertState) {
		passing = passing && (node.type === 'component' && node.status === 'alert');
	}

	if (passing) {
		return node;
	}
	else {
		const filteredChildren = node.children.map(child => filterNode(child, query, nameCheck)).filter(child => child !== null) as ITreeNode[];

		if (filteredChildren.length > 0) {
			return { ...node, children: filteredChildren };
		}
	}
	return null;
}

export function flattenTree(node: ITreeNode | null, result: ITreeNode[], query: ISearch): void {
	if (node) {
		result.push({ ...node });
		if (isCollapsed(node, query)) return;
		for (const child of node.children) {
			flattenTree(child, result, query);
		}
	}
}