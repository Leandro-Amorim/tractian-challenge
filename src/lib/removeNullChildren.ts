import { ITreeNode } from "../types/TreeNodes";

export default function removeNullChildren(tree: (ITreeNode | null)[]): ITreeNode[] {

	const filteredTree = tree.filter(node => node !== null) as ITreeNode[];

	for (const node of filteredTree) {
		node.children = removeNullChildren(node.children);
	}
	return filteredTree;
}