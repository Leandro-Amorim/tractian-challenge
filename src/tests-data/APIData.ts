import { APIAsset, APILocation } from "../types/API";

export const locations: APILocation[] = [
	{
		id: "656733b1664c41001e91d9ed",
		name: "Machinery house",
		parentId: null
	}
]

export const assets: APIAsset[] = [
	{
		id: "656734968eb037001e474d5a",
		locationId: "656733b1664c41001e91d9ed",
		name: "Motors H12D",
		parentId: null,
		sensorType: null,
		status: null
	},
	{
		id: "6567340c1f4664001f29622e",
		locationId: null,
		name: "Motor H12D- Stage 1",
		parentId: "656734968eb037001e474d5a",
		gatewayId: "QBK282",
		sensorId: "CFX848",
		sensorType: "vibration",
		status: "alert"
	},
	{
		gatewayId: "QHI640",
		id: "656734821f4664001f296973",
		locationId: "656733b1664c41001e91d9ed",
		name: "Fan",
		parentId: null,
		sensorId: "MTC052",
		sensorType: "energy",
		status: "alert"
	}
]