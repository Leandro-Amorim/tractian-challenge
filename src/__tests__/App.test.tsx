import '@testing-library/jest-dom';
import { render } from "@testing-library/react";
import App from "../App";
import buildTree from '../lib/buildTree';
import { locations, assets } from '../tests-data/APIData';
import { filterNode, flattenTree } from '../lib/filterAndFlattenTree';
import removeNullChildren from '../lib/removeNullChildren';
import { ITreeNode } from '../types/TreeNodes';

test("Renders main react component correctly", () => {
	const snapshot = render(<App />);
	expect(snapshot.asFragment()).toMatchSnapshot();
});

test('Build tree - It should build the tree correctly and return the nodes', () => {
	expect(Array.isArray(buildTree({ locations, assets }))).toBe(true);
	expect(buildTree({ locations, assets }).length).toBe(1);
	expect(buildTree({ locations, assets })[0].children.length).toBe(2);
	expect(buildTree({ locations, assets })[0].children.find(node => node.id === '656734968eb037001e474d5a')?.children.length).toBe(1);
});

test('Filter tree - It should filter the tree correctly when only the name filter is active', () => {
	const tree = buildTree({ locations, assets });

	//By location name
	let filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'machinery', energySensor: false, alertState: false, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].children.length).toBe(2);
	expect(buildTree({ locations, assets })[0].children.find(node => node.id === '656734968eb037001e474d5a')?.children.length).toBe(1);


	//By asset name
	filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'motors', energySensor: false, alertState: false, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Motors H12D');
	expect(filteredNodes[0].children[0].children.length).toBe(1);

	//By component name
	filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'stage 1', energySensor: false, alertState: false, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Motors H12D');
	expect(filteredNodes[0].children[0].children.length).toBe(1);
});


test('Filter tree - It should filter the tree correctly when only the energy sensor filter is active', () => {
	const tree = buildTree({ locations, assets });

	const filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: '', energySensor: true, alertState: false, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Fan');
});

test('Filter tree - It should filter the tree correctly when only the alert state filter is active', () => {
	const tree = buildTree({ locations, assets });

	const filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: '', energySensor: false, alertState: true, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(2);
	expect(buildTree({ locations, assets })[0].children.find(node => node.id === '656734968eb037001e474d5a')?.children.length).toBe(1);
});

test('Filter tree - It should filter the tree correctly when both the energy sensor filter and the alert state filter are active', () => {
	const tree = buildTree({ locations, assets });

	const filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: '', energySensor: true, alertState: true, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Fan');
});

test('Filter tree - It should filter the tree correctly when name, energy sensor and alert state filters are active', () => {
	const tree = buildTree({ locations, assets });

	//By location name
	let filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'machinery', energySensor: true, alertState: true, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Fan');

	//By asset name
	filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'motors', energySensor: true, alertState: true, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(0);

	//By component name
	filteredNodes = removeNullChildren(tree.map(node => filterNode(node, { query: 'fan', energySensor: true, alertState: true, collapsedIds: [] })));
	expect(filteredNodes.length).toBe(1);
	expect(filteredNodes[0].name).toBe('Machinery house');
	expect(filteredNodes[0].children.length).toBe(1);
	expect(filteredNodes[0].children[0].name).toBe('Fan');
});

test('Flatten tree - It should correctly transform the tree into a list', () => {

	const tree = buildTree({ locations, assets });
	const flattenedList: ITreeNode[] = [];

	for (const node of tree) {
		flattenTree(node, flattenedList, { query: '', energySensor: false, alertState: false, collapsedIds: [] });
	}

	expect(flattenedList.length).toBe(4);
	expect(flattenedList[0].name).toBe('Machinery house');
	expect(flattenedList[flattenedList.findIndex(node => node.id === '656734968eb037001e474d5a') + 1].name).toBe('Motor H12D- Stage 1');
});


test('Flatten tree - It should correctly hide the children of collapsed nodes', () => {
	const tree = buildTree({ locations, assets });
	const flattenedList: ITreeNode[] = [];

	for (const node of tree) {
		flattenTree(node, flattenedList, { query: '', energySensor: false, alertState: false, collapsedIds: ['656734968eb037001e474d5a'] });
	}

	expect(flattenedList.length).toBe(3);
	expect(flattenedList[0].name).toBe('Machinery house');
	expect(flattenedList.find(node => node.id === '656734968eb037001e474d5a')).toBeDefined();
	expect(flattenedList.find(node => node.name === 'Fan')).toBeDefined();
});