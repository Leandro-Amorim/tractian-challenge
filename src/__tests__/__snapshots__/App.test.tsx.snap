// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders main react component correctly 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen bg-background flex flex-col"
  >
    <header
      class="bg-header h-14 p-3 flex items-center justify-between"
    >
      <img
        alt="Tractian logo"
        class="h-5 aspect-auto hidden sm:block"
        src="./tractian.svg"
      />
      <div
        class="flex items-center gap-2 grow justify-end"
      />
    </header>
    <main
      class="grow bg-white m-[10px] rounded-lg border border-border p-4 flex flex-col"
    >
      <div
        class="flex justify-between"
      >
        <div
          class="hidden sm:flex sm:gap-2 sm:items-end"
        >
          <span
            class="text-title font-bold text-2xl"
          >
            Assets
          </span>
        </div>
        <div
          class="flex items-center gap-2 justify-end grow"
        >
          <button
            class="border rounded py-[6px] px-3 flex items-center gap-[6px] bg-white border-border"
          >
            <svg
              class="tabler-icon tabler-icon-bolt text-button-company-active"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13 3l0 7l6 0l-8 11l0 -7l-6 0l8 -11"
              />
            </svg>
            <span
              class="whitespace-nowrap font-medium text-secondary"
            >
              Energy Sensor
            </span>
          </button>
          <button
            class="border rounded py-[6px] px-3 flex items-center gap-[6px] bg-white border-border"
          >
            <svg
              class="tabler-icon tabler-icon-alert-circle text-button-company-active"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
              />
              <path
                d="M12 8v4"
              />
              <path
                d="M12 16h.01"
              />
            </svg>
            <span
              class="whitespace-nowrap font-medium text-secondary"
            >
              Alert
            </span>
          </button>
        </div>
      </div>
      <div
        class="mt-3 flex gap-4 lg:gap-2 grow flex-col lg:flex-row"
      >
        <div
          class="flex flex-col grow-0 basis-[350px] lg:grow-[3] lg:basis-0 min-w-0 border border-border rounded"
        >
          <div
            class="flex items-center border-b border-border"
          >
            <input
              class="w-full p-3 outline-none"
              placeholder="Search Asset or Location"
              value=""
            />
            <div
              class="h-full aspect-square flex items-center justify-center cursor-default"
            >
              <svg
                class="tabler-icon tabler-icon-search "
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
                />
                <path
                  d="M21 21l-6 -6"
                />
              </svg>
            </div>
          </div>
          <div
            class="flex items-center p-2 gap-2"
          >
            <button
              class="flex-1 border border-border rounded py-[6px] px-3 flex items-center justify-center"
            >
              <span
                class="font-medium text-secondary"
              >
                Collapse All
              </span>
            </button>
            <button
              class="flex-1 border border-border rounded py-[6px] px-3 flex items-center justify-center"
            >
              <span
                class="font-medium text-secondary"
              >
                Expand All
              </span>
            </button>
          </div>
          <div
            class="flex-auto"
          >
            <div
              style="overflow: visible; height: 0px; width: 0px;"
            />
            <div
              class="resize-triggers"
            >
              <div
                class="expand-trigger"
              >
                <div
                  style="width: 1px; height: 1px;"
                />
              </div>
              <div
                class="contract-trigger"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col flex-auto lg:basis-0 lg:grow-[7] min-w-0 border border-border rounded"
        />
      </div>
    </main>
  </div>
</DocumentFragment>
`;
